src/neuroglancer/datasource/computed/backend.ts

```diff
@@ -0,0 +1,617 @@
+    const k = subBoxIndex(i, /*offset=*/destLower, /*cropSize=*/commonSize, /*size=*/destSize);
+    copyFunction(j, k);
+  }
+}
+
+
+
+/**
+ * Computes a consistent key string from a chunk grid position.
+ *
+ * It's tempting to use chunk.key, in particular because these values will
+ * often be the same, but we won't always have access to a fully-specified
+ * chunk, and there's no contractual guarantee that its key will be equal to
+ * the value returned here.
+ * @param gridPosition chunk grid position
+ */
+function gridPositionKey(gridPosition: vec3) {
+  return gridPosition.toLocaleString();
+}
+
+// In addition to acting as a VolumeChunk for the purposes of a ChunkManager
+// object, also performs the book-keeping necessary to prepare the data buffer
+// used as input by the computation that provides its data. This includes
+// fetching chunk data from other datasources.
+export class ComputedVolumeChunk extends VolumeChunk implements ChunkStateListener {
+  // Defines the input and output geometry and datatypes.
+  private computationParams_?: ComputationParameters;
+
+  // Inidicates a cancellation of this chunk's computation.
+  private cancellationToken_?: CancellationToken;
+
+  // Resolve and reject functions correspond to a Promise, returned by getPromise().
+  private resolve_?: () => void;
+  private reject_?: (reason: any) => void;
+
+  // -- STATE VARIABLES: Only one of succes_, failure_, and computing_ should
+  //    resolve to true at a given time --
+
+  // If true, indicates that the computation here has been completed
+  // successfully, and data is available on this VolumeChunk.
+  private success_ = false;
+
+  // If computation failed for any reason, an indication of its cause will be
+  // stored in the failure_ variable.
+  private failure_?: any;
+
+  // True iff this chunk is actively computing.
+  private computing_ = false;
+
+  // True iff this chunk has been initialized for computation.
+  private initialized_ = false;
+
+  // A map from grid position string keys, as returned by gridPositionKey to
+  // vec3 grid positions. This is used as an indirection to avoid storing
+  // explicit references to VolumeChunks belonging to the origin source.
+  private originGridPositions_ = new Map<String, vec3>();
+
+  // The lower bound of the input patch.
+  private inputLower_?: vec3;
+
+  // Represents the input patch.
+  private inputBuffer_?: ArrayBuffer;
+
+  // Indicate to the ChunkManager that this is a computational Chunk.
+  isComputational = true;
+
+  // Our source is a ComputedVolumeChunkSource.
+  source: ComputedVolumeChunkSource;
+
+  /**
+   * Sets up computation parameters, computes overlapping origin chunks and
+   * initializes the input buffer.
+   * @param computationParams computation parameters
+   * @param cancellationToken cancellation token
+   */
+  initializeComputation(
+      computationParams: ComputationParameters, cancellationToken: CancellationToken) {
+    if (!this.source) {
```

Comment:
I think if you just create the promise here, you could avoid having the separate success_, failure_ variables and the logic for handling them, and just always directly call this.reject_ or this.resolve_.  Promise already ignores calls to reject/resolve after one has already been called so you don't have to deal with that.

Reply:
Done.